Tests/src/PublicAPITests/CancellationTokenTests.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2023 Magnopus LLC
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Awaitable.h"
+#include "CSP/CSPFoundation.h"
+#include "CSP/Common/CancellationToken.h"
+#include "TestHelpers.h"
+
+#include "gtest/gtest.h"
+
+namespace
+{
+
+#if RUN_ALL_UNIT_TESTS || RUN_CANCELLATION_TOKEN_TESTS || RUN_CANCELLATIONTOKEN_CONSTRUCTION_TEST
+CSP_PUBLIC_TEST(CSPEngine, CancellationTokenTests, ConstructionAndDestructionTest)
+{
+	csp::common::CancellationToken CancellationToken;
+	// No need to assert anything, just ensuring that construction and destruction don't crash
+}
+#endif
+
+#if RUN_ALL_UNIT_TESTS || RUN_CANCELLATION_TOKEN_TESTS || RUN_CANCELLATIONTOKEN_CANCEL_TEST
+CSP_PUBLIC_TEST(CSPEngine, CancellationTokenTests, CancelStateTest)
+{
+	csp::common::CancellationToken CancellationToken;
+	EXPECT_FALSE(CancellationToken.Cancelled());
+
+	CancellationToken.Cancel();
+	EXPECT_TRUE(CancellationToken.Cancelled());
+
+	CancellationToken.Cancel(); // Test that multiple cancellations don't affect the state
+	EXPECT_TRUE(CancellationToken.Cancelled());
+}
+#endif
+
+#if RUN_ALL_UNIT_TESTS || RUN_CANCELLATION_TOKEN_TESTS || RUN_CANCELLATIONTOKEN_COPYMOVE_TEST
+CSP_PUBLIC_TEST(CSPEngine, CancellationTokenTests, CopyMoveTest)
+{
+	// Ensure copy and move operations are deleted
+	ASSERT_FALSE(std::is_copy_constructible<csp::common::CancellationToken>::value);
+	ASSERT_FALSE(std::is_move_constructible<csp::common::CancellationToken>::value);
+	ASSERT_FALSE(std::is_copy_assignable<csp::common::CancellationToken>::value);
+	ASSERT_FALSE(std::is_move_assignable<csp::common::CancellationToken>::value);
+}
+#endif
+
+#if RUN_ALL_UNIT_TESTS || RUN_CANCELLATION_TOKEN_TESTS || RUN_CANCELLATIONTOKEN_ASYNCREF_TEST
+CSP_PUBLIC_TEST(CSPEngine, CancellationTokenTests, AsyncRefTest)
+{
+	csp::common::CancellationToken CancellationToken;
+
+	std::thread Thread(
+		[&CancellationToken]()
+		{
+			while (CancellationToken.Cancelled() == false)
+			{
+			}
```

Comment:
Adding a sleep in here would greatly reduce CPU usage during this loop. We also prefer to not explicitly compare against boolean constants. `!CancellationToken.Cancelled()` would look better here.

Also, is there any point in testing this? I can't think of a situation where holding a reference to an object in another thread would fail while said object is still alive.

Reply:
I generally agree, but also quite like it being there as a demonstration of the intended usage, as a functional test.

Reply:
Addressed in latest.